gc.c

```diff
@@ -1011,6 +1015,44 @@ asan_unlock_freelist(struct heap_page *page)
 
 #define GC_SWEEP_PAGES_FREEABLE_PER_STEP 3
 
+#define RVALUE_AGE_BITMAP_INDEX(n)  ((n) / (BITS_BITLENGTH / RVALUE_AGE_BIT_COUNT))
+#define RVALUE_AGE_BITMAP_OFFSET(n) ((n) % (BITS_BITLENGTH / RVALUE_AGE_BIT_COUNT) << 1)
+
+#define RVALUE_OLD_AGE   3
+
+static bits_t
+RVALUE_AGE_GET(VALUE obj)
+{
+    bits_t *age_bits = GET_HEAP_PAGE(obj)->age_bits;
+    return (age_bits[RVALUE_AGE_BITMAP_INDEX(NUM_IN_PAGE(obj))] >> RVALUE_AGE_BITMAP_OFFSET(NUM_IN_PAGE(obj))) & RVALUE_AGE_BIT_MASK;
+}
+
+static void RVALUE_AGE_RESET(VALUE obj);
+static void
+RVALUE_AGE_SET(VALUE obj, int age)
+{
+    RUBY_ASSERT(age <= RVALUE_OLD_AGE);
+    bits_t *age_bits = GET_HEAP_PAGE(obj)->age_bits;
+    // clear the bits
+    age_bits[RVALUE_AGE_BITMAP_INDEX(NUM_IN_PAGE(obj))] &=
+        ~(RVALUE_AGE_BIT_MASK << (NUM_IN_PAGE(obj) * RVALUE_AGE_BIT_COUNT));
```

Comment:
Don't we need to shift by `RVALUE_AGE_BITMAP_OFFSET` here? `NUM_IN_PAGE` could return a value that's much larger than 64 when it's in a separate bit plane, which will cause the left shift to result in 0.

```suggestion
        ~(RVALUE_AGE_BIT_MASK << RVALUE_AGE_BITMAP_OFFSET(NUM_IN_PAGE(obj)));
```